src/TADA_GetATTAINS.R

```diff
@@ -75,21 +63,93 @@ TADA_GetATTAINS <- function(data){
+          # remove unnecessary columns:
+          dplyr::select(-c(OBJECTID, GLOBALID)) %>%
+          # select only catchments that have WQP observations in them:
+          .[TADA_DataRetrieval_data, ] %>%
+          # prepend "ATTAINS" to ATTAINS data
+          dplyr::rename_with(~ paste0("ATTAINS.", .), dplyr::everything()) %>%
+          # get rid of dupes (as a precaution)
+          dplyr::distinct(.keep_all = TRUE),
+        silent = TRUE)
+    # if not ATTAINS data return original data frame:
+    if(is.null(nearby_catchments)){
+      print("There are no ATTAINS features associated with these WQP observations. Returning original dataframe.")
+      return(data)
+    } else {
+      # ... otherwise link WQP features to the ATTAINS catchment feature(s) they land in:
+      TADA_with_ATTAINS <- TADA_DataRetrieval_data %>%
+        sf::st_join(., nearby_catchments) %>%
+        sf::st_drop_geometry()
+      
+      if(return == TRUE){
+        
+        download_attains <- function(baseurl){
+          # bounding box (with some wiggle) of TADA data
+          bbox <- TADA_DataRetrieval_data %>% 
+            sf::st_buffer(0.001) %>% 
+            sf::st_bbox(TADA_DataRetrieval_data)
+          
+          # convert bounding box to characters
+          bbox <- toString(bbox) %>% 
+            # encode for use within the API URL
+            urltools::url_encode(.)
+          
+          # EPSG for CRS used by TADA data. THIS NEEDS TO BE IMPROVED!
+          epsg <- sf::st_crs(TADA_DataRetrieval_data)$epsg
+          
+          # set all necessary parameters for the query
+          query <- urltools::param_set(baseurl, key = "geometry", value = bbox) %>%
+            urltools::param_set(key = "inSR", value = epsg) %>%
+            urltools::param_set(key = "resultRecordCount", value = 5000) %>%
+            urltools::param_set(key = "spatialRel", value = "esriSpatialRelIntersects") %>%
+            urltools::param_set(key = "f", value = "geojson") %>%
+            urltools::param_set(key = "outFields", value = "*") %>%
+            urltools::param_set(key = "geometryType", value = "esriGeometryEnvelope") %>%
+            urltools::param_set(key = "returnGeometry", value = "true") %>%
+            urltools::param_set(key = "returnTrueCurves", value = "false") %>%
+            urltools::param_set(key = "returnIdsOnly", value = "false") %>%
+            urltools::param_set(key = "returnCountOnly", value = "false") %>%
+            urltools::param_set(key = "returnZ", value = "false") %>%
+            urltools::param_set(key = "returnM", value = "false") %>%
+            urltools::param_set(key = "returnDistinctValues", value = "false") %>%
+            urltools::param_set(key = "returnExtentOnly", value = "false") %>%
+            urltools::param_set(key = "featureEncoding", value = "esriDefault")
+          
+          final_sf <- geojsonsf::geojson_sf(query)
+          
+          return(final_sf)
+          
+        }
+        
```

Comment:
Is it possible to add more high-level explanatory comments here? I'm having trouble figuring out what's going on from here to the end of the script.